### PR TITLE
Build: Upgrade Webpack 5

### PR DESCRIPTION
**Pull request by jackw**
**What this PR does / why we need it**:

This PR does the following:

- [x] bumps Webpack to v5 and all associated plugins / loaders to their latest versions
- [x] updates all Webpack configs to work with latest Webpack / plugin / loader APIs
- [x] performance tweaks to dev mode based on [Webpack docs](https://webpack.js.org/guides/build-performance/#development)
- [x] calls `generate:theme` as a `prestart` script and removes `grafana/toolkit` from the start scripts
- [x] does the minimum required to get hot module loading working and aligns the hot config as much as possible with dev
- [x] future proofs against fork-ts-checker plugin removing Eslint support by adding `eslint-webpack-plugin` to the dependencies and configs
- [x] replaces`optimize-css-assets-webpack-plugin` with `css-minimizer-webpack-plugin` for webpack 5 compatibility
- [x] updates storybook config due to changes in SVG loader in Webpack 5 storybook config

I believe the jsIntegrity template variable in `index-template.html` isn't used but have left it in place on the off chance I'm wrong.

If you're taking the time to take this PR for a spin please remember to check the build / production output as well as dev output and storybook output. ✌️ 

---

**Special notes for your reviewer**:

<details>
  <summary>Hackathon findings (click to expand)</summary>
  
Part of my hackathon attempt was to investigate bumping webpack and all loaders / plugins to their latest versions. It updates the webpack configs inline with any breaking changes in API and attempts to make use of latest webpack features / plugins to speed up builds.

(Warning: all these timings are taken from my laptop and averaged based on three repeated runs. Probably worth taking with a pinch of sale.)

**Where are we currently?**

**Startup times:**

Version: webpack 4.41.5 Time: ~35s

**Incremental builds:**

Version: webpack 4.41.5 Time: ~8s

**What are build times like with the new version of webpack and loaders / plugins?**

**Startup times:**

Version: webpack 5.42.0 Time: ~27s

**Incremental builds:**

Version: webpack 5.42.0 Time: ~6s

**ESbuild with Webpack 5**

Modern JS bundler built on Go. Pretty much halves the time of dev builds, but it can't compile es5 so we would need to move to es2015 / es6 [which has pretty good browser support](https://caniuse.com/es6)

Biggest win speed-wise in production comes from using the esbuildMinifyPlugin instead of Terser but from my tests it appears to generate larger files. (Possibly because other tooling is still producing es5?) 🤔

The biggest issue however is the lack of babel ng-annotate plugin. Esbuild cannot decorate angular annotations which results in angular not being able to find anything once minification has occurred. Possible solution is to fork webpack-ng-annotate and get it working with webpack 5 if we want to move to Esbuild with Webpack.

Dev:
with babel - initial build ~27s | incremental builds ~6s
with esbuild - initial build ~21s | incremental builds ~3s

Prod:
with babel & terserPlugin ~134s - Entrypoints: app (5.52 MiB)
with esbuild & esbuildMinifyPlugin ~73s - Entrypoints: app (5.94 MiB)

**Webpack 5 build cache:**

Webpack 5 introduces a build cache to cache the generated webpack modules and chunks to improve build speed. Setting the type to `filesystem` could bring us some wins but I've experienced write errors during watching in dev. Enabling the build cache in dev results in cold start builds equal to incremental builds once cached.

In CI we could probably see much bigger wins by making use of the `cache.version` option and hashing the webpack configs and `.yarn-integrity` folder. 

**React Fast Refresh**

Attempting to swap out react-hot-refresh for react-fast-refresh. It appears we have some circular dependencies that are causing react-fast-refresh to error on app load. 😞 Fixed in this PR.

**94% after seal**

I've spent a lot of time trying to figure this out and I found a report that this is a combination of using eslint in Fork TS checker plugin (https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/issues/612) which blocks the type checking. However from my investigations this isn't the source of our issue.

Author of Fork TS checker plugin is [removing Eslint](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/pull/607) support in next version. So this PR brings in the `eslint-webpack-plugin` instead (which runs on a separate process).

**Federated Modules**

This caught my eye as an option for loading plugins, however it can also cater for loading different versions of packages and potentially loading frameworks only when a plugin requires them. Unfortunately I ran out of time to get a sample plugin together that demos federated module loading for plugins.

My current branches are [federated module plugin](https://github.com/jackw/grafana-plugin-federated) and [webpack5-federated](https://github.com/grafana/grafana/compare/hackathon-webpack5...hackathon-webpack5-federated) which shares react and react-dom with the plugin. 

**Webpack DLL plugin**

I'd pencilled in some time to investigate this but due time constraints I didn't get to experiment with the DLL plugin.

</details>

### CODE REVIEW

**Comment by torkelo** [comment]
YAY!!! :rocket: :tada:  :heart: 

> but it can't compile es5 so we would need to move to es2015 / es6 which has pretty good browser support

We moved to es2015/es6 in v8 with https://github.com/grafana/grafana/pull/32573

**Comment by jackw** [comment]
> We moved to es2015/es6 in v8 with #32573

Thanks so much for linking this PR @torkelo. I think it's gonna help me fix the issues with react-fast-refresh and circular dependencies. Should we update the typescript target to es6? https://github.com/grafana/tsconfig-grafana/blob/master/base.json#L25

**Comment by torkelo** [comment]
@jackw yes, wonder if that option is even used? Given we use babel in webpack, we do get ES6 classes in the build in v8 at least 🤷

**Comment by jackw** [comment]
Done here: https://github.com/grafana/tsconfig-grafana/pull/3

I've now discovered that the watchTheme command is really slowing down dev build (`yarn start`) on startup:

With watchThemes: `webpack 5.42.0 compiled successfully in 67755 ms`
Without watchThemes: `webpack 5.42.0 compiled successfully in 26023 ms`

Having disabled I've now realised it's also causing two incremental webpack builds when making changes to files. 🤔

File: public/app/features/admin/AdminSettings.tsx [hunk 1]
```diff
@@ -74,4 +72,4 @@ const mapStateToProps = (state: StoreState) => ({
   navModel: getNavModel(state.navIndex, 'server-settings'),
 });
 
-export default hot(module)(connect(mapStateToProps)(AdminSettings));
```
Comment (Clarity-89) [inline public/app/features/admin/AdminSettings.tsx:77]: Will keeping `hot(module)` break anything? We use those in Enterprise so wonder if they'd be also removed there.

**Review by Clarity-89 — COMMENTED** [review]
Tried this branch with Enterprise code and got the following errors after `yarn start`:
```
 ERROR in ../grafana-enterprise/src/public/index.ts 60:38
Module parse failed: Unexpected token (60:38)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|       {
|         path: '/datasources/edit/:id/insights',
>         component: DataSourceInsights as any,
|       },
|       {
 @ ./public/app/ sync extensions\/index.ts ./extensions/index.ts app/extensions/index.ts
 @ ./public/app/app.ts 177:24-74
 @ ./public/app/index.ts 15:0-24 16:0-8
```
```
 ERROR in public/app/extensions/meta-analytics/AnalyticsDrawer/AnalyticsDashNav.tsx:28:27
TS2604: JSX element type 'AnalyticsDrawer' does not have any construct or call signatures.
    26 |           }}
    27 |         />
  > 28 |         {isDrawerOpen && <AnalyticsDrawer dashboard={dashboard} />}
       |                           ^^^^^^^^^^^^^^^
    29 |       </>
    30 |     )
    31 |   );
```
```
ERROR in public/app/extensions/meta-analytics/AnalyticsDrawer/AnalyticsDrawer.tsx:103:14
TS2604: JSX element type 'AnalyticsUsersTab' does not have any construct or call signatures.
    101 |           )}
    102 |           {drawerTab === AnalyticsTab.Users && (
  > 103 |             <AnalyticsUsersTab dashboard={dashboard} dailySummaries={dailySummaries} userViews={userViews} />
        |              ^^^^^^^^^^^^^^^^^
    104 |           )}
    105 |         </TabContent>
    106 |       </Drawer>
```

Are we missing some extra loaders/ TS stuff here?

**Comment by jackw** [comment]
Thanks so much for taking the time to test this against enterprise @Clarity-89! 

> Are we missing some extra loaders/ TS stuff here?
I don't think it's a lack of loaders. More likely the limiting of file locations with the `include` prop. I'll take a look at it in a bit.

**Review by dprokop — COMMENTED** [review]
Took this for a spin @jackw - outstanding work!

- Core app seems to be working OK, haven't noticed any regressions (dev mode).
- Confirming @zoltanbedi observation about Icons not working in Storybook

**Comment by jackw** [comment]
Latest commits fix the following: 

- Removes production build warnings  by replacing `optimize-css-assets-webpack-plugin` with `css-minimizer-webpack-plugin` for webpack 5 compatibility.
- Fix SVG icon loading in storybook due to changes in SVG loader in Webpack 5 storybook config. There's an open PR to align iconBundle [here](https://github.com/grafana/grafana/pull/38482) that fixes the remaining missing icons.

**Comment by jackw** [comment]
> I think it would be faster if ts end eslint checks wouldn't run for test files in webpack watch.

@zoltanbedi I've tested excluding `*.test.ts`, `*.test.tsx` for typechecking and linting but I'm not seeing any speed improvements to builds as these files aren't included in our modules. What it will do is prevent builds when the test files change which kinda makes sense to me. But it raises the question where in our tooling do we rely on type and lint issues in test files? Git hooks? Drone? IDE setup?

**Review by zoltanbedi — APPROVED** [review]
Let's go!

**Pull request merged**